### PR TITLE
Library app

### PR DESCRIPTION
**Pull request by ynx0**
This PR is responsible for bringing to fruition the long-awaited library reference application.

This code is meant to be an example and will give momentum to any techniques / patterns used, (or take away from ones omitted!). Thus, I would appreciate it if you could call out any higher level changes that need to be made, in addition to general code-style/best practices.

### CODE REVIEW

File: lib/library.hoon [hunk 1]
```diff
@@ -0,0 +1,135 @@
+/-  *post, resource, library, spider
+/+  *graph-store, strandio
+|%
+++  incr-index
+  :: increments the last value in the index by 1
+  |=  [=index:post]
+  ^-  index:post
+  =/  last=@                (sub (lent index) 1)  :: calculate the array index of the last item
+  =/  old-revision-count=@  (snag last index)     :: get the last item
```
Comment (tacryt-socryp) [inline lib/library.hoon]: Can use `+rear` instead.

File: mar/graph/validator/library.hoon [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+      i
+    ::  single metadata revision
+    ::  content node. first %text element is treated as title,
+    ::  second is treated as isbn
+    ::
+        [@ %meta @ ~]
+      =/  contents  contents.p.i
+      ?>  ?=([[%text *] [%text *] ~] contents.p.i)
+      =/  title  (trip +.i.contents.p.i)
+      ?>  (is-title-valid title)
```
Comment (tacryt-socryp) [inline mar/graph/validator/library.hoon:45]: Why does this need a tape rather than a cord? Same question for ISBN

File: mar/graph/validator/library.hoon [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    ::
+        [@ %comments @ ~]
+      :: we could allow any content eventually
+      ::?>  ?=(* contents.p.i)  :: any content is allowed.
+      ?>  ?=([[%text *] ~] contents.p.i)  :: only a single %text content is allowed
+      =/  contents  contents.p.i
+      i
+    ==
+  ::  vestigial structures from graph-push-hook. don't know if we still need these
+  ++  graph-permissions-add
```
Comment (tacryt-socryp) [inline mar/graph/validator/library.hoon]: No need

File: mar/graph/validator/library.hoon [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    |=  vip=vip-metadata:met
+    ^-  permissions:graph
+    !!
+  ::
+  ++  graph-permissions-remove
```
Comment (tacryt-socryp) [inline mar/graph/validator/library.hoon]: No need

File: mar/graph/validator/library.hoon [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    |=  vip=vip-metadata:met
+    ^-  permissions:graph
+    !!
+  ++  transform-add-nodes
```
Comment (tacryt-socryp) [inline mar/graph/validator/library.hoon]: If you aren't using graph-push-hook, you don't need this. (Unless you are using it in your proxy-hook!)

**Review by tacryt-socryp — COMMENTED** [review]
- [ ] Will do a more thorough review, but let me get this straight: a library is a graph, and you can set individual permissions per book in the library, or set permissions by library? Is there any reason you're constraining the use case to just text, or? I imagine the networking logic here will be generally useful for many other applications (email, etc). Would love to learn more on your intentions for how this can be used.

**Comment by ynx0** [comment]
> * [ ]  Will do a more thorough review, but let me get this straight: 
> a library is a graph

yes
> and you can set individual permissions per book in the library, or set permissions by library? 

TL;DR: Per library. 
All access control is handled by the policy, thus anyone who is granted access to the library (a successful watch in the `%request-library` flow) then has *access* to all books, but is not tracking updates to any of them yet. Then, once the user requests to start tracking a book in the form of `%request-book` poke, they will now get the book's node and start to receive updates for it. Put another way, `%request-library` is where a subscriber is accepted/rejected by the hose, while `%request-book` is the subscriber choosing which book he/she wants to start tracking (i.e. see it's contents and any comments). 

> Is there any reason you're constraining the use case to just text, or? I imagine the networking logic here will be generally useful for many other applications (email, etc). Would love to learn more on your intentions for how this can be used.

Yes, the hope is definitely that networking logic here is generic enough to provide a (slightly opinionated) scaffolding for others to build their applications. However, this is one specific example / instantiation of a social media application built on top of `%graph-store`, rather than being a generic pattern / fill in the blank application template. The reason I used text for title/isbn was just because it in my eyes the goal of the project was more for showing how to do the networking logic while keeping the design of the schema relatively simple. (It also kept the project more tractable for me, since I had little prior Hoon experience and basically learned it on the job, as you can tell from the code). This is also why the application is in more of a PoC state in terms of true utility to a user. In its current state, you are really only storing a reference to a book which maybe the frontend would populate a nice image, and using it to pass not-even-nested commentary. So if you were to hook the app up to a nice frontend, the user would probably find it cute at best, rather than a compelling application to use store books. 

In general, the original aim was the simplest possible social media app that would both show off core features of graph store while simultaneously allowing others to copy and adapt to ease the process of creating their own social media apps. It could definitely be improved and taken to its full potential, i.e. a epub/pdf/book reader where you can actually store the contents of the book and have users highlight/annotate it in a shared environment, but again, that was not really the goal of this specific proposal.

File: lib/library.hoon (hunk 1 above)
Comment (ynx0) [reply lib/library.hoon]: So I could replace `(snag last index)` with `(rear index)`, but in the call to `+snap` I'd still need `last`, right? Would you know of a better way to replace the tail of the index so we could get rid of it altogether?

File: mar/graph/validator/library.hoon (hunk 2 above)
Comment (ynx0) [reply mar/graph/validator/library.hoon:45]: Hmmm. tbh I didn't put too much thought into it so its probably just a not so good decision. Looking at `cord`, it supports UTF-8 and `tape` doesn't so thats one more argument against it. I think the main reason was because I didn't know/forgot how to get the length of a `cord`. For the length check for the isbn. Once I know how to do that I'll change both to cords.

File: mar/graph/validator/library.hoon (hunk 5 above)
Comment (ynx0) [reply mar/graph/validator/library.hoon]: Well, as of now I'm not using it, but it does bring up the question of if we *should* be using it. IIRC this arm was basically an extra step that would ensure all time based indexes of an incoming post were modified to be when they were received on the host, as opposed to trusting the author. (Although I don't know exactly how it works). If this API is here to stay, and it's a best practice in your eyes, then I can try to figure out which places it would make sense to run an update through the arm.

File: mar/graph/validator/library.hoon (hunk 4 above)
Comment (ynx0) [reply mar/graph/validator/library.hoon]: addressed in https://github.com/ynx0/library/pull/3/commits/650cc1046585d83cfc71774f6c70503a48ffb4e5

File: mar/graph/validator/library.hoon (hunk 3 above)
Comment (ynx0) [reply mar/graph/validator/library.hoon]: addressed in https://github.com/ynx0/library/pull/3/commits/650cc1046585d83cfc71774f6c70503a48ffb4e5

File: sur/library.hoon [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
++$  comment  tape
++$  book
+  $:
+    title=tape
+    isbn=tape
+    :: TODO fulltext=tape ? 
+  ==
+::
+::  TODO better / clearer name for prim
++$  prim     (jug resource atom)  :: for a given resource, what indexes does the reader want to hear about?
```
Comment (tacryt-socryp) [inline sur/library.hoon]: add a comment that this is a networking section

File: sur/library.hoon [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
++$  readers  (map ship prim)      :: given a ship, what are the resources and books that it cares about?
+::  TODO most things that take a rid can theoretically be just a name=@tas because entity=@p is always gonna be the owner of the proxy
```
Comment (tacryt-socryp) [inline sur/library.hoon:16]: yes remove the redundancy

File: mar/graph/validator/library.hoon (hunk 5 above)
Comment (ynx0) [reply mar/graph/validator/library.hoon]: OOB: Logan: complexity of transform add nodes is not justified for this example, we don't need its functionality anyways, so its good to remove

File: lib/library.hoon (hunk 1 above)
Comment (tacryt-socryp) [reply lib/library.hoon]: `(snoc (snip index) +((rear index)))`

File: lib/library.hoon [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+++  index-to-path
+  |=  [idx=index:post]
+  ^-  path
+  (turn idx (cury scot %ud))
+::
+::
+++  create-library-update
+  |=  [[owner=ship name=@tas] time-sent=time]
+  ^-  update
+  :-  time-sent
```
Comment (tacryt-socryp) [inline lib/library.hoon:40]:
```
:^  time-sent  %add-graph
  [owner name]
[*graph `%graph-validator-library %.n]
```

same thing, just play around with the style and see what you like best!

File: lib/library.hoon [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+      [~ %graph-validator-library]
+      %.n     :: don't overwrite an existing library
+  ==
+++  remove-library-update
+  |=  [[owner=ship name=@tas] time-sent=time]
+  ^-  update
+  :-  time-sent
+  :-  %remove-graph
+  [owner name]
+++  add-book-update
```
Comment (tacryt-socryp) [inline lib/library.hoon]: scry out the whole graph, run `(pry:orm graph)` and then you'll have the most recent element

Comment (tacryt-socryp) [reply lib/library.hoon]: this applies to `revise-book-update`, not this arm

File: app/library-proxy.hoon [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+  ^-  (quip card _this)
+  ~&  >  '%library-proxy recompiled successfully'
+  `this(state !<(versioned-state old-state))
+++  on-poke
+  |=  [=mark =vase]
+  ^-  (quip card _this)
+  =^  cards  state
+  ?+    mark  (on-poke:def mark vase)
+      %library-command
+    ::?>  (team:title our.bowl src.bowl)  :: allow ourselves and moons to use this poke
```
Comment (tacryt-socryp) [inline app/library-proxy.hoon]: I think we're moving away from moons acting as equivalent to the planet in terms of permissions for 3rd party software distribution, probably get rid of this

File: lib/library.hoon (hunk 9 above)
Comment (ynx0) [reply lib/library.hoon]: https://github.com/ynx0/library/pull/3/commits/54b2fc3c9317f13687c0fc8e3d93945a020a4dde

File: mar/graph/validator/library.hoon (hunk 2 above)
Comment (ynx0) [reply mar/graph/validator/library.hoon:45]: ~~ok thinking about it more, isbn should only ever be numbers and dashes so it makes sense for it to remain a tape. not title, however.~~

edit: nvm, just looked at what tape actually is, which is a list of chars, which are already utf8 code points anyways. gonna make everything a cord now

Comment (ynx0) [reply mar/graph/validator/library.hoon:45]: https://github.com/ynx0/library/pull/3/commits/f86bedbaefebcc85541a15b7c411499369e4c680

File: sur/library.hoon (hunk 7 above)
Comment (ynx0) [reply sur/library.hoon:16]: https://github.com/ynx0/library/pull/3/commits/243ca90a3e12420a464436c3a35012097c670d70

File: app/library-proxy.hoon (hunk 10 above)
Comment (ynx0) [reply app/library-proxy.hoon]: https://github.com/ynx0/library/pull/3/commits/1fdca233683c347287800451016bfae4128ba125

File: lib/library.hoon (hunk 8 above)
Comment (ynx0) [reply lib/library.hoon:40]: https://github.com/ynx0/library/pull/3/commits/e0e660826361bf85fd8beda15a64217c9fc82b35

File: sur/library.hoon (hunk 6 above)
Comment (ynx0) [reply sur/library.hoon]: https://github.com/ynx0/library/pull/3/commits/e0e660826361bf85fd8beda15a64217c9fc82b35

File: lib/library.hoon (hunk 1 above)
Comment (ynx0) [reply lib/library.hoon]: https://github.com/ynx0/library/pull/3/commits/3035efda739f00f164a40ec78364fc16101b8913

File: mar/graph/validator/library.hoon (hunk 5 above)
Comment (ynx0) [reply mar/graph/validator/library.hoon]: https://github.com/ynx0/library/pull/3/commits/bccc429ef0d6c182f720201d5c5a7a8d31610019

File: lib/library.hoon (hunk 8 above)
Comment (ynx0) [reply lib/library.hoon:40]: https://github.com/ynx0/library/pull/3/commits/86f60a7f7ed52771d7bbc945b9557a397b810224

**Review by tacryt-socryp — APPROVED** [review]
LGTM
